src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,117 @@
+
+            context.RegisterRefactoring(
+                new ConvertAutoPropertyToFullPropertyCodeAction(
+                    FeaturesResources.Convert_to_full_property,
+                    c => ExpandToFullPropertyAsync(document, property, propertySymbol, root, cancellationToken)));
+        }
+
+        internal bool IsValidAutoProperty(SyntaxNode property, IPropertySymbol propertySymbol)
+        {
+            var fields = propertySymbol.ContainingType.GetMembers().OfType<IFieldSymbol>();
+            var field = fields.FirstOrDefault(f => propertySymbol.Equals(f.AssociatedSymbol));
+            return field != null;
+        }
+
+        private async Task<Document> ExpandToFullPropertyAsync(
+            Document document,
+            SyntaxNode property,
+            IPropertySymbol propertySymbol,
+            SyntaxNode root,
+            CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var workspace = document.Project.Solution.Workspace;
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            // Create full property. If the auto property had an initial value
+            // we need to remove it and later add it to the backing field
+            var fieldName = await GetFieldNameAsync(document, propertySymbol, cancellationToken).ConfigureAwait(false);
+            var accessorTuple = GetNewAccessors(options, property, fieldName, generator);
+            var fullProperty = generator
+                .WithAccessorDeclarations(
+                    GetPropertyWithoutInitializer(property),
+                    accessorTuple.newSetAccessor == null
+                        ? new SyntaxNode[] { accessorTuple.newGetAccessor }
+                        : new SyntaxNode[] { accessorTuple.newGetAccessor, accessorTuple.newSetAccessor })
+                .WithLeadingTrivia(property.GetLeadingTrivia());
+            fullProperty = ConvertPropertyToExpressionBodyIfDesired(options, fullProperty);
+            var editor = new SyntaxEditor(root, workspace);
+            editor.ReplaceNode(property, fullProperty.WithAdditionalAnnotations(Formatter.Annotation));
+
+            // add backing field, plus initializer if it exists 
+            var newField = CodeGenerationSymbolFactory.CreateFieldSymbol(
+                default, Accessibility.Private,
+                DeclarationModifiers.From(propertySymbol),
+                propertySymbol.Type, fieldName,
+                initializer: GetInitializerValue(property));
+            var containingType = GetTypeBlock(
+                propertySymbol.ContainingType.DeclaringSyntaxReferences.FirstOrDefault().GetSyntax(cancellationToken));
+            editor.ReplaceNode(containingType, (currentTypeDecl, _)
+                => CodeGenerator.AddFieldDeclaration(currentTypeDecl, newField, workspace)
+                .WithAdditionalAnnotations(Formatter.Annotation));
```

Comment:
this is very heavyweight.  you're essentialy formatting the *entire* type declaration.  is that necessary?